src/source/tcp.rs

```diff
@@ -134,44 +139,54 @@ where
             let mut events = mio::Events::with_capacity(1024);
             match poll.poll(&mut events, None) {
                 Err(e) => panic!(format!("Failed during poll {:?}", e)),
-                Ok(_num_events) => {
-                    for event in events {
-                        match event.token() {
-                            constants::SYSTEM => {
-                                for handler in self.handlers {
-                                    handler.shutdown();
-                                }
-
-                                util::send(&mut chans, metric::Event::Shutdown);
-                                return;
+                Ok(_num_events) => for event in events {
+                    match event.token() {
+                        constants::SYSTEM => {
+                            for handler in self.handlers {
+                                handler.shutdown();
                             }
-                            listener_token => {
-                                self.spawn_stream_handlers(
-                                    chans.clone(), // TODO: do not clone, make an Arc
-                                    tags,
-                                    listener_token,
-                                );
+
+                            util::send(&mut chans, metric::Event::Shutdown);
+                            return;
+                        }
+                        listener_token => {
+                            let listener = &(self.listeners[listener_token]);
+                            match self.spawn_stream_handlers(
+                                chans.clone(),
+                                tags,
+                                listener,
+                            ) {
+                                Ok(new_handlers) => {
+                                    self.handlers.extend(new_handlers);
```

Comment:
Why does `spawn_stream_handlers` allocate a Vec for the new handlers if we're subsequently extending that vec onto `self.handlers`? I guess I'm missing something but it seems like we could avoid the allocation and keep tacking on to `self.handlers`?

Reply:
There was an issue with immutable vs mutable borrows resulting from - https://github.com/postmates/cernan/pull/398/files/e04f650864a0da3a7fded0f7eeb13a09bd1ece63#diff-c657fbced7b730476f584243df38b7baR153.  That is what prompted me to have this function return a Vec.

I can pass in a ref to self.handlers instead.  Is there another way to resolve the conflict?

Reply:
Hmm, let me fiddle some. Mind if I push a commit here? 

FWIW there's changing coming down the pipe in the borrow checker that will make these cases much less common.